### PR TITLE
[WFLY-19954] Add the vertx extension/subsystem from wildfly-vertx-feature-pack to WildFly Preview Feature Pack

### LINKED ISSUE #662: Add the vertx extension/subsystem from wildfly-vertx-feature-pack to WildFly Preview Feature Pack
### Description

There is a discussion in [smallrye-reactive-messaging to configure vertx options](https://github.com/smallrye/smallrye-reactive-messaging/discussions/2725) to ask the capability to configure the vertx options used in the microprofile-reactive-messaging subsystem, and there is a [wildfly-vertx-feature-pack](https://github.com/wildfly-extras/wildfly-vertx-feature-pack) ready to fulfill the requirement, integrate it to WildFly should satisfy the requirement.

More discussions can be found at:
* https://wildfly.zulipchat.com/#narrow/channel/174184-wildfly-developers/topic/vertx.20subsystem.20integration
* https://lists.jboss.org/archives/list/wildfly-dev@lists.jboss.org/thread/CHUJ5GZSHXUSVZNXS3X32H7UQ4M64NKC/

### Issue Contact

gaol

### PR DESCRIPTION
**Pull request by gaol**
Issue: https://issues.redhat.com/browse/WFLY-19954

Resolve: https://github.com/wildfly/wildfly-proposals/issues/662

### CODE REVIEW

**Comment by gaol** [comment]
@kabir @jasondlee Would you please review ? thanks.

**Comment by bstansberry** [comment]
Hi @gaol -- FYI re https://github.com/wildfly/wildfly/pull/18466. If that solves the WFLY-20035 problem the vertx extension will likely need similar handling.

File: testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Tests that it should fail to inject a Vertx instance with a not-existed qualifier provided by vertx subsystem.
+     */
+    @Test
+    public void testFail() throws Exception {
+        try {
+            HttpRequest.get(url.toExternalForm() + "rest/echo/fail/Hello", 4, TimeUnit.SECONDS);
+            Assert.fail("Not here");
```
Comment (kabir) [inline testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:75]: s/Not here/Should have failed

**Review by kabir — COMMENTED** [review]
Hi @gaol 

I'm not super-familiar with CDI, so I just wanted to double check what happens in the `!exist` case in this EchoService method:

```
private Vertx getVertx(boolean exist) {
        if (exist) {
            return vertxInstance.select(Identifier.Literal.of("vertx")).get();
        }
        return vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get();
    }
```

My guess is that `vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get()` throws some kind of exception, and thus the GET in `VertxSubsystemSmokeTestCase.testFail()` fails ?

It looks like that reading the code, just wanted to double check :-)

File: testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java (hunk 1 above)
Comment (gaol) [reply testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:75]: > My guess is that `vertxInstance.select(Identifier.Literal.of("vertx-not-exist")).get()` throws some kind of exception, and thus the GET in `VertxSubsystemSmokeTestCase.testFail()` fails ?

Yes, it is, the CDI exception(`org.jboss.weld.exceptions.UnsatisfiedResolutionException`
) gets propagated to client via `return Response.status(500).entity(e).build()` and should contain 
 a specific message which is checked in the test:

```java
Assert.assertTrue(e.getMessage().contains("WELD-001334: Unsatisfied dependencies for type Vertx"));
```

**Comment by gaol** [comment]
> Hi @gaol -- FYI re #18466. If that solves the WFLY-20035 problem the vertx extension will likely need similar handling.

yes, I created https://github.com/wildfly-extras/wildfly-vertx-feature-pack/issues/46 to track the changes in the vertx extension and will update this one later after another release.

File: testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * Tests that it should fail to inject a Vertx instance with a not-existed qualifier provided by vertx subsystem.
+     */
+    @Test
+    public void testFail() throws Exception {
+        try {
+            HttpRequest.get(url.toExternalForm() + "rest/echo/fail/Hello", 4, TimeUnit.SECONDS);
+            Assert.fail("Should have failed");
+        } catch (IOException e) {
+            Assert.assertTrue(e.getMessage().contains("WELD-001334: Unsatisfied dependencies for type Vertx"));
```
Comment (kabir) [inline testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:77]: Actually, here I guess this message could be internationalised, so it is probably better to just check for "WELD-001334"

Comment (gaol) [reply testsuite/preview/basic/src/test/java/org/wildfly/test/preview/vertx/VertxSubsystemSmokeTestCase.java:77]: Ah, yes. thank you. will update it  :)

**Comment by gaol** [comment]
Thank you @bstansberry @kabir for the comments,  I updated this PR with:
* Upgrade the wildfly-vertx-feature-pack to 1.0.2.Final with https://github.com/wildfly-extras/wildfly-vertx-feature-pack/issues/46 addressed
* Update the test code to check the exception id only.

**Review by kabir — APPROVED** [review]

**Comment by bstansberry** [comment]
@jasondlee @fl4via Please review

**Review by jasondlee — APPROVED** [review]
lgtm

**Comment by gaol** [comment]
As this PR has conflicts and smallrye-opentelemetry has been upgraded to 2.9.0, I updated this PR with additional commits:

* Fixed code conflicts
* Added `io.smallrye.common.annotation` to the dependency of `io.smallrye.opentelemetry` module
* Put `otel.exporter.vertx.cdi.identifier` with value of `vertx` to OpenTelemetryConfig
* Put `mp.messaging.connector.vertx.cdi.identifier` with value of `vertx` to ReactiveMessagingConfigSource

The latter 2 changes make the 2 subsystems try to lookup the Vertx instance from the CDI identifier, if there is no one exposed from vertx subsystem, it falls back to what they currently do(call Vertx.vertx())

**Review by jasondlee — APPROVED** [review]

**Review by kabir — APPROVED** [review]

**Review by fl4via — APPROVED** [review]

**Pull request merged**